### ISSUE TITLE
Error during processing of JPG -- macOS Catalina 10.15

### ISSUE THREAD

**Issue by robskrob** [tesseract-ocr/tesseract#2930]
### Environment

* **Tesseract Version**: 4.1.1
* **Platform**: macOS Catalina 10.15

### Current Behavior:
Whenever I execute `$ tesseract images/IMG_3958.jpeg output/grocery_bill` I get this error:
```
$ tesseract images/IMG_3958.jpg output/grocery_bill
Tesseract Open Source OCR Engine v4.1.1 with Leptonica
Error during processing.
```

### Expected Behavior:
I would expect `tesseract` to output the text from the grocery bill into the output file `output/grocery_bill`.

Is there something wrong with the image? Also, is there a location where I can tail the logs to see if I can get a richer description of the error?

Here's more information about the `tesseract` program that I installed with Homebrew:
```
$ tesseract -v
tesseract 4.1.1
 leptonica-1.79.0
  libgif 5.2.1 : libjpeg 9d : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 1.1.0 : libopenjp2 2.3.1
 Found AVX2
 Found AVX
 Found FMA
 Found SSE
```

Also attached please find the image I had `tesseract` process.
![IMG_3958](https://user-images.githubusercontent.com/4349424/77260327-8eb1d980-6c5d-11ea-882c-7df920b73f1c.jpg)

**Comment by stweil** [comment]
You closed that, so what was the solution?

**Comment by robskrob** [comment]
@stweil I apologize for closing without an explanation. 

The problem is the image is not a JPG. It is a HEIC. When take a photo with my iphone, upload it to google drive and then download this photo onto my machine, the image is saved as a HEIC file, which to be honest is a new file extension to me.

So the "solution" for me was to convert the file on my machine from HEIC to JPG. Once I did this `tesseract` had no problem processing the image. 

I suppose then my original issue is still a bug -- except there's nothing wrong with the JPG file. It's the HEIC file:
```
$ tesseract images/IMG_3958.HEIC output/grocery_bill
Tesseract Open Source OCR Engine v4.1.1 with Leptonica
Error during processing.
```

Attached please find the HEIC file

[IMG_3958.HEIC.zip](https://github.com/tesseract-ocr/tesseract/files/4366376/IMG_3958.HEIC.zip)

**Comment by stweil** [comment]
Thanks. So the error message should be improved and report that the input file could not be read.

**Comment by robskrob** [comment]
Yeah I wanted to tail the logs of my `tesseract` process so I could potentially learn more about what was going. I definitely think the error message could be improved. And yes, there's something about the input file that appears to not be readable. And yes, chancing the title makes sense.

**Comment by stweil** [comment]
See the [Wikipedia article on HEIC](https://en.wikipedia.org/wiki/High_Efficiency_Image_File_Format). Maybe Leptonica can be extended to support that new format. There exists a C library [libheif](https://github.com/strukturag/libheif) which might be used. @DanBloomberg, would that be interesting?

**Comment by DanBloomberg** [comment]
I've asked an expert on coding about this format, and will report here when I hear back.

For leptonica to support it is a high bar, because support is a serious commitment and invariably a lot of work.  That work will include passing very intensive fuzzing, both internally at Google and externally by oss-fuzz on github.

**Comment by baryluk** [comment]
It is more of a political / maintainability decision than technical one, but leptonica already supports webp, so maybe heic / heif should be considered too? AFAIK Canon EOX  1Dx Mk III support natively writing to HEIF instead of JPEG, which is kind of a big deal and big commitment from Canon already.

**Comment by amitdo** [comment]
Maybe Leptonica/Tesseract can support gdk-pixbuf, which will bring indirect support for heic and avif.

* https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/issues/119
* https://gitlab.gnome.org/GNOME/eog/-/issues/36
* https://gitlab.gnome.org/GNOME/gthumb/-/issues/88
* https://github.com/strukturag/libheif#gdk-pixbuf-loader
* https://developer.gnome.org/gdk-pixbuf/
* https://packages.debian.org/bullseye/libgdk-pixbuf-2.0-0
* https://packages.debian.org/bullseye/heif-gdk-pixbuf
* https://packages.debian.org/bullseye/libavif-gdk-pixbuf

**Comment by DanBloomberg** [comment]
Each new format supported demands a big cost in development and support effort.

As mentioned on leptonica#546, jpeg-xl seems more interesting: both easier to support and having compatibility with jpeg.

**Comment by amitdo** [comment]
Dan. I understand your view.

The thing is, these two formats (HEIC and AVIF) are becoming quite popular. HEIC is the default file format for photos in iOS. Recently Firefox followed Chrome to support AVIF by default (previously users had to enable it manually).

So, since you do not want to support these formats in your software. maybe we, Tesseract devs, should discuss whether we want to support it ourselves.

If we will decide to support them. we will convert them to Leptonica's pix.

To be honest, this was mainly directed toward @stweil, hoping that he would want this enough to implement it... :-)

**Comment by DanBloomberg** [comment]
Supporting AVIF and HEIC within tesseract is certainly an option.  As I said, my experience with the older I/O libraries is that it's a lot of work.  webp has been easier, because there are not a lot of options and the implementation is more "modern", with the basic encoder and decoder going between memory and not file streams (or, worse, with tiff, unix file descriptors).  With the fuzzers that have been made for leptonica, both internally in Google and now externally with the oss-fuzz project, maintenance work has increased considerably. 

But overall, maintenance on the I/O libraries has been a significant time-sink.  I'm glad to have been able to relieve Ray Smith and tesseract of that burden.

I can't promise anything about jpeg-xl, but it does seem to be something I should look into.

**Comment by amitdo** [comment]
>I'm glad to have been able to relieve Ray Smith and tesseract of that burden.

Last time we saw Ray here was 2 years ago. I don't know if he plans to return to contribute to this project.

**Comment by amitdo** [comment]
>So, since you do not want to support these formats in your software. maybe we, Tesseract devs, should discuss whether we want to support them ourselves.

Answering myself: We should not do it.

**Comment by amitdo** [comment]
Still, a proper error message for unsupported image formats is desired.

**Comment by DanBloomberg** [comment]
pixReadStream() emits the message:
    Unknown format: no pix returned
if the format is not supported for reading.

I believe that tesseract should not rely on leptonica error messages -- you may not even emit them by default.
Instead, tesseract should have its own error message if the image file can't be read to a pix.

**Comment by zdenop** [comment]
The funny part is that OCR process of this image fails here:
https://github.com/tesseract-ocr/tesseract/blob/b5878c23a70a6709e722aea5a3304a4c5c87313b/src/api/baseapi.cpp#L974-L976

Because  when the file format is unknown (`IFF_UNKNOWN`), tesseract API(???) expects it is a file list:

https://github.com/tesseract-ocr/tesseract/blob/b5878c23a70a6709e722aea5a3304a4c5c87313b/src/api/baseapi.cpp#L1201-L1210

I think we should remove this "guess what is the input" out of OCR API...

**Comment by vid-bin** [comment]
Is there any update on this? My entire photo library is heic (including screenshots). I’ve been trying to get OCR working with heic on fscrawler and Nextcloud with no success.

OCR works on my Mac because Apple has support for it in their system but I want to get this working on my NAS.

the only solution would be to convert my library to jpeg which I do not want to do.

**Comment by amitdo** [comment]
We don't plan to support the HEIC format.

**Comment by stweil** [comment]
No, there isn't any update, neither for Leptonica nor for Tesseract. The only solution is currently to convert from HEIC to JPEG, so Tesseract can process the JPEG file.

**Comment by vid-bin** [comment]
Does tesseract support AVIF or jpeg-xl then? I don’t want to convert to jpeg because I’ll lose HDR on my photos. The storage space is also significantly higher when converting my images.

Ideally the applications indexing the files would convert to jpeg on the fly for tesseract and then delete the temporary file when done but the ones I’m trying to use do not do this.

**Comment by DanBloomberg** [comment]
We've looked a couple of times at jpeg-xl.  There is a very high bar to cross before deciding to support a new format with leptonica, and the reasons for this have been described several times.  AVIF  doesn't meet it.  Initially, jpeg-xl was an interesting possibility because it was supported by Google, it is a significantly more efficient encoder than jpeg, and it is designed to have very good compatibility with the jpeg library.  All these things meant that there was some likelihood of widespread adoption within a few years, essentially as a replacement for jpeg, and the work of supporting it in leptonica would be less than that for a completely new format.

Just a reminder: part of the work to support any new format is to build and run fuzzers for months, and to harden it not to crash or hang for any possible input.

As of 2022, however, it was evident that Google was not supporting it.
For details why it is no longer of interest, see:
     https://github.com/DanBloomberg/leptonica/issues/692#issuecomment-1501259293

**Comment by amitdo** [comment]
https://tesseract-ocr.github.io/tessdoc/InputFormats

**Comment by DanBloomberg** [comment]
For the tesseract input formats page, you might clarify that reading animated webp (a-webp) is not supported -- only writing.

I lost interest in fully supporting a-webp when I found that Google had so little interest in supporting their own format (which is far superior in compression to animated gif) that even with several billion gmail accounts, they didn't let you view an a-webp attachment!  Only the first frame is displayed.  Nevertheless, a-webp is supported in browsers; see the Google webp faq[https://developers.google.com/speed/webp/faq](url)

**Comment by amitdo** [comment]
Hi Dan,

Thanks for the info. Is the description on that page in the 'Animated GIF' section also correct for animated WebP?

**Comment by DanBloomberg** [comment]
No, leptonica can not read a webp anim file.  It does not return the first image.

Here are the error messages:
```
Error in pixReadMemWebP: WebP decode failed
Error in pixReadStream: webp: no pix returned
```
